Keep track of aliased indices

martijnvg opened the pull request:
The PR adds a top level `aliasedIndices` map to `Metadata` that keeps track of all the aliases that are defined in each `IndexMetadata`. This `aliasedIndices` map isn't rebuild on each `Metadata#build(...)` invocation, but rather updated in place each time alias definitions are added or removed from `IndexMetadata` instances. New `Metadata.Builder` instances reuse the previous `aliasedIndices` map.

This change is beneficial because it will allow the `indicesLookup` sorted map to always be build lazily (which is what this PR is also doing). Prior to this change building the `indicesLookup` sorted map was required for certain alias validation, but this is no longer the case, since that validation can now use `aliasedIndices` map. Building the `indicesLookup` sorted map can get expensive in clusters with many indices and always being able to delay this computation to after a new cluster state has been build and published, can reduce load on the cluster state update thread on the elected master node.

Also two helper methods have been added to `Metadata` class to easily and efficiently get access to aliases and the indices these aliases refer to. Logically aliases are often seen as independent resources, but this isn't the case and getting access to an entire alias requires checking all `IndexMetadata` instances. This change will make this easier, without changing how we store aliases in the cluster state. Changing that would be much more difficult.

martijnvg committed:
first commit

joegallo committed:
Add getters for aliases

joegallo committed:
Increase consistency

getIndices() and getAliases() go through indices() and aliases(), so
do that here, too.

joegallo committed:
Failing unit test

joegallo committed:
And make it pass

joegallo committed:
Test the newly created alias, too

joegallo committed:
Failing unit test

joegallo committed:
And make it pass

joegallo committed:
Clean up imports / use of liberally

joegallo committed:
Test for deleting an index with an alias

joegallo committed:
Failing unit test

joegallo committed:
And make it pass

martijnvg committed:
Merge remote-tracking branch 'es/master' into refactor_indices_aliases

martijnvg committed:
add AliasIndicesReference that encapsulates the set of indices an alias refers to

joegallo committed:
Merge remote-tracking branch 'martijnvg/refactor_indices_aliases' into refactor_indices_aliases

joegallo committed:
Keep this immutable

joegallo committed:
Check arguments

martijnvg committed:
fix tests

joegallo committed:
Rename this method

to better reflect that it's "given an alias, tell me which indices
points to"

joegallo committed:
Reimplement hasAlias in terms of aliases

joegallo committed:
Oops, fix a bad test

I'm sure this must have been intended to use the previously created
random request rather than ignoring that one entirely and creating a
brand new one.

joegallo committed:
Drop this helper method

It's odd to have a request class poking through here, and then all we
do is unwrap it to get the underlying array of aliases. Better IMHO to
make the array-consuming version public and have callers do the
unwrapping themselves.

joegallo committed:
Rewrite getIndicesLookup slightly

and move the unmodifiable wrapping into buildIndicesLookup itself

joegallo committed:
Prefer hasIndexAbstraction over containsKey

getIndicesLookup().containsKey() is just hasIndexAbstraction() -- it
does the exact same thing internally, but the intent reads a little
clearer.

martijnvg committed:
A number of changes:
* Update aliasesIndices map aliasesIndices when updating IndexMetadata instance on Metadata.Builder.
* Also no need to serialize aliasesIndices map anymore, since it gets aliasesIndices build.
* Move alias write indices validation from IndexAbstraction.Alias to Metadata.Builder#build()

martijnvg committed:
rename aliases fields in Metadata and Metadata.Builder class to aliasesIndices

martijnvg committed:
rename aliasesIndices to aliasedIndices

martijnvg committed:
undo unneeded change

martijnvg committed:
Use a setting that is actually allowed to be changed.

Also this breaks validation logic in Metadata.Builder.removeAlias() method,
we keep track of the uuid of an index in aliasedIndices map, so changing
this would also imply updating that. But this setting is immutable and
can't be changed after an index has been created. (just in tests we seem
to be able to change this setting)

martijnvg committed:
spotless

joegallo committed:
Rewrite this test in terms of Metadata building

joegallo committed:
Move validateAliasProperties into Metadata

Combines validateAliasProperties and validateAlias into a
rewritten (and relocated) validateAlias. The Index->IndexMetadata
lookup is outside the validation to make it a little more obvious that
that's happening.

martijnvg committed:
Removed builtIndicesLookupEagerly functionality

martijnvg committed:
turn validateDataStreams into assertion

joegallo committed:
spotless

joegallo committed:
Merge branch 'master' into refactor_indices_aliases

martijnvg committed:
method rename

joegallo committed:
Don't need this comment anymore

joegallo committed:
Merge branch 'master' into refactor_indices_aliases

martijnvg committed:
assert no aliases refer to backing indices of data streams by checking
that IndexMetadata of these backing indices have no AliasMetadata defined.

martijnvg committed:
removed unused getter

martijnvg committed:
added unit test for aliased indices and
added java docs

martijnvg committed:
improve unit test and a method to retrieve the name of all aliases

joegallo committed:
YAGNI the AliasIndicesReference

We can always bring it back later if we do end up wanting it.

joegallo committed:
Move these alias tests to MetadataTests

joegallo committed:
Merge branch 'master' into refactor_indices_aliases

elasticmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

martijnvg commented on server/src/test/java/org/elasticsearch/cluster/metadata/MetadataTests.java:
```diff
@@ -1294,51 +1296,7 @@ public void testValidateDataStreamsAllowsNamesThatStartsWithPrefix() {
-        validateDataStreams(metadata.getIndicesLookup(), (DataStreamMetadata) metadata.customs().get(DataStreamMetadata.TYPE));
-    }
-
-    public void testValidateDataStreamsAllowsPrefixedBackingIndices() {
```
This test was testing whether certain backing index naming was allowed.
Only data stream naming convention no longer cares about this.
If during a rollover a backing index by accident (which would be very rare) collides with an existing index then 
we have logic that deals with this by increasing generation (which would generate another backing index name).

@danhermann Do you think that this (^) assessment is correct? In other words this test is obsolete and can be removed?

danhermann replied:
Yes, I think that this test has been superseded by DataStreamTests::testRolloverWithConflictingBackingIndexName.

martijnvg committed:
Merge remote-tracking branch 'es/master' into refactor_indices_aliases

martijnvg committed:
oops

martijnvg committed:
Two changes:
* Metadata#hasAlias(...) should also check for data stream aliases.
  (The contract of this method doesn't make a difference between index aliases and data stream aliases)
* Don't invoke getIndicesLookup() when comparing whether indices lookup is different between two Metadata instances.
  Added a new method that can compare indices lookups for assertions that doesn't invoke the getIndicesLookup() method.
  In cases when indices lookup hasn't been invoked then there may not yet be an indices lookup and then null will be
  compared against an indices lookup (may be empty), which would always make assertions trip.

martijnvg committed:
Merge remote-tracking branch 'es/master' into refactor_indices_aliases

martijnvg committed:
Removed unused imports

martijnvg committed:
Merge remote-tracking branch 'es/master' into refactor_indices_aliases

martijnvg committed:
Merge remote-tracking branch 'es/master' into refactor_indices_aliases

original-brownbear commented on server/src/internalClusterTest/java/org/elasticsearch/cluster/ClusterStateDiffIT.java:
```diff
@@ -570,9 +570,7 @@ public IndexMetadata randomChange(IndexMetadata part) {
-                        builder.settings(
-                            Settings.builder().put(part.getSettings()).put(IndexMetadata.SETTING_INDEX_UUID, UUIDs.randomBase64UUID())
-                        );
+                        builder.settings(Settings.builder().put(part.getSettings()).put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0));
```
Why do we need this change?

martijnvg replied:
Changing the UUID of an index isn't possible in real life, but this test does change it and
it fails validation logic that was added in `Metadata.Builder.removeAlias()` method.

In this method we keep track of aliased indices by `Index` and that contains the uuid.
The exception on line 1359 is thrown. Basically because during this test, we (randomly) 
add an index with alias, change the uuid setting and then remove the alias. An entry
is attempted be removed, but doesn't exist, because uuid changed.

original-brownbear replied:
Ah got it, so our validation actually got better here I guess :)

martijnvg replied:
I guess so :) 

All of our settings validation occurs when providing settings updates via the api. In tests, like here all the settings validation is by-passed, because we directly work with `Settings` class.

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:
```diff
@@ -1267,6 +1303,70 @@ public Builder indices(ImmutableOpenMap<String, IndexMetadata> indices) {
+            Objects.requireNonNull(alias);
+            Objects.requireNonNull(index);
+
+            Objects.requireNonNull(alias);
```
These two lines seem accidentally duplicated

original-brownbear reviewed (commented):
Looks good :) Just one question on why we need to change that one test and one cosmetic thing.

Just to note this here for others not involved in this work: this PR is a massive speedup to building the metadata on master nodes. In many-shards benchmarking it led to up to a 50% increase in the rate of index creation in large clusters. The larger the cluster the more helpful this change is :)
=> thanks so much Martijn!

martijnvg committed:
Merge remote-tracking branch 'es/master' into refactor_indices_aliases

martijnvg committed:
removed duplicated checks

original-brownbear reviewed (approved):
LGTM

martijnvg commented:
Thanks for reviewing @original-brownbear!

martijnvg commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into refactor_indices_aliases
